azurelinuxagent/ga/policy/policy_engine.py

```diff
@@ -41,15 +65,118 @@ def _log_policy(cls, msg, is_success=True, op=WALAEventOperation.Policy, send_ev
+        Policy will be enabled if (1) policy file exists at CUSTOM_POLICY_PATH and (2) the conf flag "Debug.EnableExtensionPolicy" is true.
         Caller function should check this before performing any operations.
         """
-        # TODO: Add check for policy file present at /etc/waagent_policy.json.
         # Policy should only be enabled if conf flag is true AND policy file is present.
-        return conf.get_extension_policy_enabled()
+        policy_file_exists = os.path.exists(CUSTOM_POLICY_PATH)
+        return conf.get_extension_policy_enabled() and policy_file_exists
+
+    def __get_policy(self):
+        """
+         Check if custom policy exists at CUSTOM_POLICY_PATH, load JSON object and return as a dict.
+        Return {} if no policy exists.
+        The expected policy format is:
+         {
+            "policyVersion": str,
+            "extensionPolicies": {
+                "allowListedExtensionsOnly": bool,
+                "signatureRequired": bool,
+                "extensions": {
+                    "<extensionName1>": {
+                        "signatureRequired": bool
+                    }
+                }
+            }
+        }
+        """
+        # TODO: Add schema validation for custom policy file and raise relevant error message to user.
+        if os.path.exists(CUSTOM_POLICY_PATH):
+            self._log_policy("Custom policy found at {0}. Using custom policy.".format(CUSTOM_POLICY_PATH))
+            with open(CUSTOM_POLICY_PATH, 'r') as f:        # Open file in read-only mode.
+                custom_policy = json.load(f)
+                return custom_policy
+        else:
+            self._log_policy("No custom policy found at {0}. Using default policy.".format(CUSTOM_POLICY_PATH))
+            return {}
+
+
+class ExtensionPolicyEngine(_PolicyEngine):
+
+    def __init__(self, extension_to_check):
+        """
+        Extension_to_check is expected to be an Extension object (main Plugin/handler specified by the publishers).
+        Each instance of ExtensionPolicyEngine is tied to an extension.
+        """
+        super(ExtensionPolicyEngine, self).__init__()
+        self._extension_to_check = extension_to_check
+        if not self.is_policy_enforcement_enabled():
+            return
+
+        extension_policy = self.policy.get("extensionPolicies", {})
+        # CRP allows extension names to be any case. We use a case-folded dict to do a case-insensitive query of the allowlist.
+        # The rest of the policy remains case-sensitive, so we only replace the "extensions" dict with a case-folded dict,
+        # if it exists.
+        if extension_policy.get("extensions") is not None:
+            case_folded_extension_dict = _CaseFoldedDict.from_dict(extension_policy.get("extensions"))
+            self.policy["extensionPolicies"]["extensions"].update(case_folded_extension_dict)
+
+        self.extension_policy = extension_policy
+
+    def should_allow_extension(self):
```

Comment:
Will this function will be called for every ext, in every goal state? Even if policy is not enabled?

Reply:
This function will be called any time we try to enable an extension (right before the handle_enable call in handle_ext_handler()). It will be called even if policy isn't enabled, but if policy is not enabled, it will just return True.